File: app/src/main/java/maderski/markparkingspot/LocationManager.java
package maderski.markparkingspot;

import android.Manifest;
import android.content.Context;
import android.content.pm.PackageManager;
import android.location.Location;
import android.util.Log;
import android.widget.Toast;

import java.util.Calendar;
import java.util.List;

/**
 * Created by Jason on 2/7/16.
 */
public class LocationManager extends LocationUpdater{
    private static final String TAG = LocationManager.class.getName();

    private Location locationResult;

    //Get location
    public LocationManager(Context context){
        super(context);
        checkPermissions(context);
    }

    private void checkPermissions(Context context){
        //Get current GPS and Network permissions
        PackageManager packageManager = context.getPackageManager();
        int hasGPSPermission = packageManager.checkPermission(Manifest.permission.ACCESS_FINE_LOCATION,
                context.getPackageName());
        int hasNetworkPermission = packageManager.checkPermission(Manifest.permission.ACCESS_COARSE_LOCATION,
                context.getPackageName());
        //Check if Permission is granted
        if(hasGPSPermission == PackageManager.PERMISSION_GRANTED && hasNetworkPermission == PackageManager.PERMISSION_GRANTED) {
            locationManager = (android.location.LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
            //locationResult = locationManager.getLastKnownLocation(LocationManager.GPS_PROVIDER);
        }else {
            if(BuildConfig.DEBUG)
                Log.e(TAG, "Need GPS Permission!!!");
            Toast.makeText(context, "Need GPS Permission!", Toast.LENGTH_LONG).show();
        }
    }

    private void listAllProviders(){
        List<String> matchingProviders = locationManager.getAllProviders();
        for(String provider: matchingProviders){
            Log.i("CurrentLocation", " Provider: " + provider);
            Log.i("CurrentLocation", " Accuracy: " + Float.toString(locationResult.getAccuracy()));
            Log.i("CurrentLocation", " Time: " + Long.toString(locationResult.getTime()));
        }
    }

    public void getBestLocation(Context context){
        if(gpsLock != null){
            if(!isAccuracyPoor(gpsLock)) {
                locationResult = gpsLock;
                if (BuildConfig.DEBUG) {
                    Log.i(TAG, "GPS LOCK USED: " + gpsLock.getAccuracy());
                    Toast.makeText(context, "GPS", Toast.LENGTH_SHORT).show();
                }
                return;
            }
        } else if(netLastLock != null) {
            if(!isAccuracyPoor(netLock)) {
                locationResult = netLock;
                if (BuildConfig.DEBUG) {
                    Log.i(TAG, "NETWORK LOCK USED: " + netLock.getAccuracy());
                    Toast.makeText(context, "NETWORK", Toast.LENGTH_SHORT).show();
                }
                return;
            }
        } else {
            useLaskKnowLocation(context);
            if(BuildConfig.DEBUG){
                Log.i(TAG, "LAST KNOWN LOCATION USED");
            }
        }
    }

    private void useLaskKnowLocation(Context context){
        getLastKnownLocation(context);
        if(gpsLastLock.getTime() > netLastLock.getTime()) {
            if(gpsLastLock.getTime() > Calendar.getInstance().getTimeInMillis() - 120000) {
                locationResult = gpsLastLock;
                if(BuildConfig.DEBUG) {
                    Toast.makeText(context, "GPS LAST", Toast.LENGTH_SHORT).show();
                }
            }
        }
        else{
            if(netLastLock.getTime() > Calendar.getInstance().getTimeInMillis() - 120000) {
                locationResult = netLastLock;
                if(BuildConfig.DEBUG) {
                    Toast.makeText(context, "NET LAST", Toast.LENGTH_SHORT).show();
                }
            }
        }
    }

    //Returns true if GPS is enabled
    public boolean isGPSAndNetworkEnabled(){
        return locationManager.isProviderEnabled(android.location.LocationManager.GPS_PROVIDER) &&
                locationManager.isProviderEnabled(android.location.LocationManager.NETWORK_PROVIDER);
    }

    //Return Latitude as a String
    public String getLatitude(){
        String latitude = "None";
        try {
            if (locationManager.isProviderEnabled(android.location.LocationManager.GPS_PROVIDER)) {
                if(locationResult != null)
                    latitude = Double.toString(locationResult.getLatitude());
            }
        }catch (Exception e){
            Log.e(TAG, e.getMessage());
        }
        return latitude;
    }

    //Return Longitude as a String
    public String getLongitude(){
        String longitude = "None";
        try {
            if (locationManager.isProviderEnabled(android.location.LocationManager.GPS_PROVIDER))
                if(locationResult != null)
                    longitude = Double.toString(locationResult.getLongitude());
        }catch(Exception e){
            Log.e(TAG, e.getMessage());
        }
        return longitude;
    }

    //Return Accuracy as a string
    public String getAccuracyAsString(){
        String Accuracy = "unknown";

        try{
            if(locationManager.isProviderEnabled(android.location.LocationManager.GPS_PROVIDER)){
                if(locationResult != null) {
                    double locationToFeet = locationResult.getAccuracy() * 3.28;
                    Accuracy = Double.toString(locationToFeet);
                }
            }
        }catch (Exception e){ e.printStackTrace(); }

        return Accuracy;
    }

    public boolean isAccuracyPoor(Location location){
        float Accuracy = 30;
        try{
            if(locationManager.isProviderEnabled(android.location.LocationManager.GPS_PROVIDER) &&
                    locationManager.isProviderEnabled(android.location.LocationManager.NETWORK_PROVIDER)){
                Accuracy = location.getAccuracy();
            }
        }catch (Exception e){ e.printStackTrace(); }

        if(Accuracy > 16.0){ return true; }
        else{ return false; }
    }

    public Location getLocationResult(){
        return locationResult;
    }
}

File: app/src/main/java/maderski/markparkingspot/LocationManager.kt
package maderski.markparkingspot

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.location.Location
import android.util.Log
import android.widget.Toast
import android.location.LocationManager

import java.util.Calendar

/**
 * Created by Jason on 2/7/16.
 */
class LocationManager(context: Context) : LocationUpdater(context) {

    var locationResult: Location? = null
        private set

    //Returns true if GPS is enabled
    val isGPSAndNetworkEnabled: Boolean
        get() = locationManager.isProviderEnabled(android.location.LocationManager.GPS_PROVIDER) && locationManager.isProviderEnabled(android.location.LocationManager.NETWORK_PROVIDER)

    //Return Latitude as a String
    val latitude: String
        get() {
            var latitude = "None"
            try {
                if (locationManager.isProviderEnabled(android.location.LocationManager.GPS_PROVIDER)) {
                    locationResult?.let {
                        latitude = java.lang.Double.toString(it.latitude)
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, e.message)
            }

            return latitude
        }

    //Return Longitude as a String
    val longitude: String
        get() {
            var longitude = "None"
            try {
                val isGPSEnabled = locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)
                if (isGPSEnabled) {
                    locationResult?.let {
                        longitude = it.longitude.toString()
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, e.message)
            }

            return longitude
        }

    //Return Accuracy as a string
    val accuracyAsString: String
        get() {
            return try {
                val isGPSEnabled = locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)
                if (isGPSEnabled) {
                    var accuracy = "unknown"
                    locationResult?.let {
                        val locationToFeet = it.accuracy * 3.28
                        accuracy = locationToFeet.toString()
                    }
                    accuracy
                } else {
                    "unknown"
                }
            } catch (e: Exception) {
                e.printStackTrace()
                "unknown"
            }
        }

    init {
        checkPermissions(context)
    }

    private fun checkPermissions(context: Context) {
        //Get current GPS and Network permissions
        val packageManager = context.packageManager
        val hasGPSPermission = packageManager.checkPermission(Manifest.permission.ACCESS_FINE_LOCATION,
                context.packageName)
        val hasNetworkPermission = packageManager.checkPermission(Manifest.permission.ACCESS_COARSE_LOCATION,
                context.packageName)
        //Check if Permission is granted
        if (hasGPSPermission == PackageManager.PERMISSION_GRANTED && hasNetworkPermission == PackageManager.PERMISSION_GRANTED) {
            locationManager = context.getSystemService(Context.LOCATION_SERVICE) as android.location.LocationManager
        } else {
            if (BuildConfig.DEBUG)
                Log.e(TAG, "Need GPS Permission!!!")
            Toast.makeText(context, "Need GPS Permission!", Toast.LENGTH_LONG).show()
        }
    }

    private fun listAllProviders() {
        val matchingProviders = locationManager.allProviders
        for (provider in matchingProviders) {
            locationResult?.let {
                Log.i("CurrentLocation", " Provider: $provider")
                Log.i("CurrentLocation", " Accuracy: " + java.lang.Float.toString(it.accuracy))
                Log.i("CurrentLocation", " Time: " + java.lang.Long.toString(it.time))
            }
        }
    }

    fun getBestLocation(context: Context) {
        if (gpsLock != null) {
            val isAccuracyGood = isAccuracyPoor(gpsLock as Location).not()
            if (isAccuracyGood) {
                locationResult = gpsLock
                if (BuildConfig.DEBUG) {
                    Log.i(TAG, "GPS LOCK USED: " + gpsLock?.accuracy)
                    Toast.makeText(context, "GPS", Toast.LENGTH_SHORT).show()
                }
                return
            }
        } else if (netLastLock != null && netLock != null) {
            val isAccuracyGood = isAccuracyPoor(netLock as Location).not()
            if (isAccuracyGood) {
                locationResult = netLock
                if (BuildConfig.DEBUG) {
                    Log.i(TAG, "NETWORK LOCK USED: " + netLock?.accuracy)
                    Toast.makeText(context, "NETWORK", Toast.LENGTH_SHORT).show()
                }
                return
            }
        } else {
            useLaskKnowLocation(context)
            if (BuildConfig.DEBUG) {
                Log.i(TAG, "LAST KNOWN LOCATION USED")
            }
        }
    }

    private fun useLaskKnowLocation(context: Context) {
        getLastKnownLocation(context)
        if (gpsLastLock != null && gpsLastLock != null) {
            val gpsLastLockTime = (gpsLastLock as Location).time
            val netLastLockTime = (netLastLock as Location).time
            if (gpsLastLockTime > netLastLockTime) {
                if (gpsLastLockTime > Calendar.getInstance().timeInMillis - 120000) {
                    locationResult = gpsLastLock
                }
            } else {
                if (netLastLockTime > Calendar.getInstance().timeInMillis - 120000) {
                    locationResult = netLastLock
                }
            }
        }
    }

    fun isAccuracyPoor(location: Location): Boolean {
        var accuracy = 31f
        try {
            val isGPSEnabled = locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)
            val isNetworkEnabled = locationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER)
            if (isGPSEnabled && isNetworkEnabled) {
                accuracy = location.accuracy
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return accuracy > 30.0f
    }

    companion object {
        private const val TAG = "LocationManager"
    }
}
